File: src/test/java/de/flapdoodle/generics/uniontypes/j/FieldGeneratorTest.java
package de.flapdoodle.generics.uniontypes.j;

import de.flapdoodle.generics.uniontypes.j.components.DateField;
import de.flapdoodle.generics.uniontypes.j.components.FieldFactory;
import de.flapdoodle.generics.uniontypes.j.components.TextField;
import de.flapdoodle.generics.uniontypes.j.framework.Component;
import de.flapdoodle.generics.uniontypes.j.framework.HasValidation;
import de.flapdoodle.generics.uniontypes.j.framework.HasValue;
import de.flapdoodle.generics.uniontypes.j.model.Property;
import kotlin.Suppress;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Supplier;

import static org.junit.jupiter.api.Assertions.*;

class FieldGeneratorTest {
  @Test
  void fieldGeneratorUseCase() {
    FieldGenerator generator = new FieldGenerator(new DummyFieldFactory());
    Container container = new Container();

    generator.addFields(container, listOf(new Property<>(String.class, "One")));

    assertEquals(1, container.children.size());
    assertEquals(TextField.class, container.children.get(0).getClass());
  }

  private <T> List<T> listOf(T...values) {
    return Arrays.asList(values);
  }

  class Container extends Component {
  private List<Component> children = new ArrayList<>();
    public void add(Component component) {
      children.add(component);
    }
  }

  class DummyFieldFactory implements FieldFactory {
    @Override
    public <T, C extends Component & HasValue<T> & HasValidation> Supplier<C>
    factoryFor(Class<T> type) {
      if (type == String.class) {
        return () -> (C) new TextField();
      }
      if (type == LocalDate.class) {
        return () -> (C) new DateField();
      }
      throw new IllegalArgumentException("unknown $type");
    }
  }
}

File: src/test/kotlin/de/flapdoodle/generics/uniontypes/k/FieldGeneratorTest.kt
package de.flapdoodle.generics.uniontypes.k

import de.flapdoodle.generics.uniontypes.k.components.DateField
import de.flapdoodle.generics.uniontypes.k.components.FieldFactory
import de.flapdoodle.generics.uniontypes.k.components.TextField
import de.flapdoodle.generics.uniontypes.k.framework.Component
import de.flapdoodle.generics.uniontypes.k.framework.HasValidation
import de.flapdoodle.generics.uniontypes.k.framework.HasValue
import de.flapdoodle.generics.uniontypes.k.model.Property
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.time.LocalDate

internal class FieldGeneratorTest {

  @Test
  fun fieldGeneratorUseCase() {
    val generator = FieldGenerator(DummyFieldFactory())
    val container = Container()

    generator.addFields(container, listOf(Property(String::class.java, "One")))

    assertEquals(1, container.children.size)
    assertEquals(TextField::class.java, container.children.get(0)::class.java)
  }

  class Container(val children: MutableList<Component> = mutableListOf()) : Component() {
    override fun add(component: Component) {
      children.add(component)
    }
  }

  @Suppress("UNCHECKED_CAST")
  class DummyFieldFactory : FieldFactory {
    override fun <T, C> factoryFor(type: Class<out T>): () -> C
        where C : Component, C : HasValue<T>, C: HasValidation {
      if (type == String::class.java) {
        return { TextField() as C }
      }
      if (type == LocalDate::class.java) {
        return { DateField() as C }
      }
      throw IllegalArgumentException("unknown $type")
    }

  }
}